pkg/types/intoto/v0.0.2/entry.go

```diff
@@ -0,0 +1,418 @@
+		}
+		// Not all in-toto statements will contain a SLSA provenance predicate.
+		// See https://github.com/in-toto/attestation/blob/main/spec/README.md#predicate
+		// for other predicates.
+		if predicate, err := parseSlsaPredicate(decodedPayload); err == nil {
+			if predicate.Predicate.Materials != nil {
+				for _, s := range predicate.Predicate.Materials {
+					for alg, ds := range s.Digest {
+						result = append(result, alg+":"+ds)
+					}
+				}
+			}
+		}
+	default:
+		log.Logger.Infof("Unknown in_toto DSSE envelope Type: %s", *v.IntotoObj.Content.Envelope.PayloadType)
+	}
+	return result, nil
+}
+
+func parseStatement(p []byte) (*in_toto.Statement, error) {
+	ps := in_toto.Statement{}
+	if err := json.Unmarshal(p, &ps); err != nil {
+		return nil, err
+	}
+	return &ps, nil
+}
+
+func parseSlsaPredicate(p []byte) (*in_toto.ProvenanceStatement, error) {
+	predicate := in_toto.ProvenanceStatement{}
+	if err := json.Unmarshal(p, &predicate); err != nil {
+		return nil, err
+	}
+	return &predicate, nil
+}
+
+func (v *V002Entry) Unmarshal(pe models.ProposedEntry) error {
+	it, ok := pe.(*models.Intoto)
+	if !ok {
+		return errors.New("cannot unmarshal non Intoto v0.0.2 type")
+	}
+
+	var err error
+	if err := types.DecodeEntry(it.Spec, &v.IntotoObj); err != nil {
+		return err
+	}
+
+	// field validation
+	if err := v.IntotoObj.Validate(strfmt.Default); err != nil {
+		return err
+	}
+
+	if string(*v.IntotoObj.Content.Envelope.Payload) == "" {
+		return errors.New("DSSE envelope does not contain a payload")
+	}
+
+	env := &dsse.Envelope{
+		Payload:     string(*v.IntotoObj.Content.Envelope.Payload),
+		PayloadType: *v.IntotoObj.Content.Envelope.PayloadType,
+	}
+
+	allPubKeyBytes := make([][]byte, 0)
+	for _, sig := range v.IntotoObj.Content.Envelope.Signatures {
+		env.Signatures = append(env.Signatures, dsse.Signature{
+			KeyID: sig.Keyid,
+			Sig:   string(sig.Sig),
+		})
+
+		allPubKeyBytes = append(allPubKeyBytes, sig.PublicKey)
+	}
+
+	if _, err := verifyEnvelope(allPubKeyBytes, env); err != nil {
+		return err
+	}
+
+	v.env = *env
+
+	decodedPayload, err := base64.StdEncoding.DecodeString(string(*v.IntotoObj.Content.Envelope.Payload))
+	if err != nil {
+		return fmt.Errorf("could not decode envelope payload: %w", err)
+	}
+
+	h := sha256.Sum256(decodedPayload)
+	v.IntotoObj.Content.PayloadHash = &models.IntotoV002SchemaContentPayloadHash{
+		Algorithm: swag.String(models.IntotoV002SchemaContentPayloadHashAlgorithmSha256),
+		Value:     swag.String(hex.EncodeToString(h[:])),
+	}
+
+	return nil
+}
+
+func (v *V002Entry) Canonicalize(ctx context.Context) ([]byte, error) {
+
+	canonicalEntry := models.IntotoV002Schema{
+		Content: &models.IntotoV002SchemaContent{
+			Envelope: &models.IntotoV002SchemaContentEnvelope{
+				PayloadType: v.IntotoObj.Content.Envelope.PayloadType,
+				Signatures:  v.IntotoObj.Content.Envelope.Signatures,
+			},
+			Hash:        v.IntotoObj.Content.Hash,
+			PayloadHash: v.IntotoObj.Content.PayloadHash,
```

Comment:
why not include the payloadhash inside the envelope? So in the envelope scheme there is currently a payload field, and maybe it can be extended to have either a payload content or a payloadHash. That way users can support adding verifiable attestations with hidden/private payload content.

Reply:
Should this be a large discussion on if we want to handle this now or in the future? I understand the use case but has the community adopted this yet?

Reply:
I think we haven't been able to adopt it yet because there wasn't support -- since it's just a minor change to schema, allowing users to upload and set the payload hash would be nice to do in this iteration.

But on the other hand that does change `verifyEnvelope` -- and the impl https://github.com/secure-systems-lab/go-securesystemslib/blob/main/dsse/verify.go here would need to change. 

Would movidying a `readOnly` attribute constitute a breaking API change? @bobcallaway 

But regardless, moving the payloadhash inside the payload field wouldn't change anything and I think it looks organizationally better to group the payload attributes

Reply:
why should users be able to upload the payload hash? I think we should compute it server-side only. If people only want to provide a hash, then they should use `hashedrekord` only.

Reply:
>  If people only want to provide a hash, then they should use hashedrekord only.

This is fair, closing this. There is only one benefit: the multi-sig properties of intoto. But don't get any benefit from the type-specific information that intoto gives you over hashedrekord.